aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -409,10 +437,17 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
       dstBuffer.add_(srcBuffer, value);
     }
   } else {
+    Tensor promoted_result = r;
+    if (r.scalar_type() != commonDtype) {
+      promoted_result = r.to(commonDtype);
+    }
     AT_DISPATCH_ALL_TYPES(
-        values.scalar_type(), "add_dense_sparse", [&] {
-          add_dense_sparse_worker_cpu<scalar_t>(r, value, sparse, indices, values);
+        commonDtype, "add_dense_sparse", [&] {
+          add_dense_sparse_worker_cpu<scalar_t>(promoted_result, value, sparse, indices, values);
         });
+    if (r.scalar_type() != commonDtype) {
+      r.copy_(promoted_result);
+    }
```

Comment:
The implementation is not the nicest; it looks like to do type promotion, the current implementation adds the logic to each possible pathway. Instead of doing that, we could imagine writing a wrapper function around the implementation that promotes the input sparse tensors to the correct dtype, calls the implementation (that operates on tensors of the same dtype), and then copy the result back to the result tensor.

Reply:
yeah I'd like to clean this up eventually, but in doing so I think I'd want to extend support for more combinations of tensors too. Right now arithmetic ops on sparse tensors are pretty limited and didn't want to try to make it generic without actually making the availability of sparse ops more generic.

Reply:
I don't see why "adding more support for sparse ops" and "making the implementation more generic" have to go together. Right now, the implementation for `add_sparse_out_cpu` is hard to review because there are two code paths that need to be considered, one for where values are contiguous, and one for when they're not.

It would make it easier to review (and also make the implementation look better) if we did something like the following (pseudocode)

```
add_out_sparse_cpu(result, tensor, other) {
  // do type promotion
  common_dtype = result_type(tensor, other)
  result_tmp = empty_like(result, common_dtype)
  tensor_tmp = tensor.to(common_dtype)
  other_tmp = other.to(common_dtype)

  // call the base op
  add_out_sparse_cpu_no_type_promotion(result_tmp, other_tmp, tensor_tmp)

  // shove the result back into its original type.
  result.set_inplace(result_tmp.indices(), result_tmp.values().to(result.dtype), result_tmp.sizes())
}
```

Reply:
In general putting the type promotion code in helper functions or a wrapper function like this would help because the body of `add_out_sparse_cpu` as well as the other functions are really, really long